### DISCUSSION TITLE
Modifying EB_UserDefined.H

### DISCUSSION THREAD

**Discussion by Mawssen** [AMReX-Combustion/PeleLMeX#346] [General]
Hello Everyone,
I'm Massen, I've just started working with PeleLMeX. I'm in the process of running and customizing templates to get familiar with the environment. Currently, I'm running the "Backward facing step anchored premixed flame" template, and I aim to simply add an extra step to the upper part of the box. Despite several attempts in editing EBUserDefined.H, I couldn't get what I want (nothing shows up!). Could you guide me on how to add another step into my simulation?
You can see the sketch below, that this is the step I want to add to my simulation:

![2024-02-20 22 31 58](https://github.com/AMReX-Combustion/PeleLMeX/assets/119066584/43dd25a0-9f5b-4d1b-ad0a-17d9c1a732e1)

This is my code in EB_UserDefined.H:

void
EBUserDefined(
  const Geometry& geom,
  const int required_coarsening_level,
  const int max_coarsening_level)
{
  // ParmParse your geometry parameters
    
  // Build geometry pieces using EB2::* methods
  EB2::BoxIF box({-0.01, 0.005}, {0.01, 0.010},false);

  // Build your geometry shop using EB2::makeShop
  auto gshop = EB2::makeShop(box);

  // Build geom using EB2::Build
  EB2::Build(gshop, geom, 0,0);
}

**Comment by marchdf** [comment]
This might help, might not. In our other code, PeleC, we do things slightly differently but this guts of the code looks the same and might give you lots of examples of geometries: https://github.com/AMReX-Combustion/PeleC/blob/development/Source/Geometry.cpp

The geometry you have there _should_ have added a single box as you defined it. It will however be the only thing in the domain (if your domain bounds are set right). Also I am pretty sure you want `EB2::Build(gshop, geom, max_coarsening_level, max_coarsening_level);` Also put some print statements in there for now to make sure you actually enter that function ;)

**Comment by baperry2** [comment]
Yeah this capability is definitely under-documented. The examples @marchdf points out should be helpful to show the capability. The functions for building and combining different EB shapes are all defined in the AMReX source code, so you can look in there for more information as well. But what you want to do is pretty simple and can be accomplished by combining two boxes:

```
  EB2::BoxIF box1({-0.01, -0.01}, {0.01, 0.00},false);
  EB2::BoxIF box2({-0.01, 0.005}, {0.01, 0.01},false);
  auto total_geom = EB2::makeUnion(box1,box2);
  auto gshop = EB2::makeShop(total_geom);
  EB2::Build(gshop, geom, required_coarsening_level, max_coarsening_level);
```

Also, note that due to a file naming convention change, this function should go in `PeleLMeX_EBUserDefined.H`. Finally, remember to set `eb2.geom_type= UserDefined` in your input file.

**Reply by Mawssen** [reply]
Thank you for the solution that worked! Now, I am attempting to simulate the classic problem of fully developing channel flow. The initial condition involves zero velocity across the entire domain at t=0, with an inflow of velocity U_inf=5 gradually developing within the channel until it reaches a fully developed state. Could you recommend a template to modify for achieving this? I attempted to modify the EB_FlowPastCylinder template by removing the cylinder and adding walls to the top and bottom. In the pelelmex_prob.H file, I set the initial velocity values to zero as follows:
  
```
pelelmex_initdata(...)
{
switch (prob_parm.meanFlowDir) {
  case 1:
    AMREX_D_TERM(state(i, j, k, VELX) = 0.0;
                 , state(i, j, k, VELY) = 0.0;, state(i, j, k, VELZ) = 0.0);
...
```
However, this modification does not seem to have any impact on the simulation. I also attempted to adjust the input.2d-Re500 file by changing values for prob.meanFlowMag, but this did not yield the desired results. Can you please guide me on the simplest way to simulate this problem and provide information on how to control boundary conditions and initial velocity data?

Thank you!

**Reply by esclapez** [reply]
The first thing LMeX does is perform an initial velocity projection, which basically propagates the uniform velocity you impose at the boundary through the entire domain to ensure mass conservation. This overwrites your uniform zero-velocity field in this case (if you had velocity perturbations such as vortices in initdata, those would still be present on top of the mean velocity field imposed by the boundary condition).
So if you really want to see a zero velocity initial solution you can either de-activate the initial projection or have a temporal ramp up of your velocity in the bcnormal function. However, there is nothing wrong with the post initial projection velocity field.

**Reply by Mawssen** [reply]
Thank you for your guidance. Could you please provide instructions on how to deactivate the initial velocity projection or how to add the post-initial projection velocity field to enforce my desired initial condition? because I tried a couple of things but still have the same issue!

**Reply by esclapez** [reply]
So, if you want to deactivate the initial steps taken by LMeX:
```
peleLM.num_divu_iter = 0
peleLM.num_init_iter = 0
peleLM.do_init_proj = 0
```
Note that this will allow you to have your initdata velocity field in the `plt00000`, but as soon as LMeX performs a time step, a projection occurs and thus your boundary condition on velocity will impose a 'mean' velocity to conserve mass. I am not sure what you are trying to achieve, but an initial condition consistent with your boundary conditions is only available after the projection. Having a zero-velocity initial field with a non-zero velocity on the boundary does not satisfy the incompressible constraint.
